### PR TITLE
Make SpectatorClient thread-safe

### PR DESCRIPTION
**Pull request by smoogipoo**
Prereqs:
- [x] https://github.com/ppy/osu/pull/12881

`SpectatorClient` not being thread-safe leads to quite a lot of issues:
- We have to ensure proper locking and scheduling in other parts of code, such as `SpectatorScreen`, which can't schedule itself since sometimes a subscreen is pushed to it (and thus it becomes unable to receive updates).
    - This can also lead to deadlocks as I've experienced, especially if we're to expose bindables such as `PlayingUsers` and eventually the user states directly.
    - It also isn't done correctly _already_. For example, the replay is mutated directly on the SignalR thread, showcasing how easy this is to get wrong:  
https://github.com/ppy/osu/blob/b5ef4b608b073bd5ba63d1f8889ed63c9cf29be4/osu.Game/Screens/Spectate/SpectatorScreen.cs#L192-L201
- We can't do things such as exiting the `SpectatorScreen` when a user restarts gameplay, unless `SpectatorScreen` schedules to the `Game` anyway (which actually isn't currently possible since we don't expose `Schedule()` on game/arbitrary Drawables). This is currently handled by duplicate code in two places:  
https://github.com/ppy/osu/blob/b5ef4b608b073bd5ba63d1f8889ed63c9cf29be4/osu.Game/Screens/Play/SpectatorPlayer.cs#L73-L81
https://github.com/ppy/osu/blob/b5ef4b608b073bd5ba63d1f8889ed63c9cf29be4/osu.Game/Screens/Play/SpectatorResultsScreen.cs#L28-L37

The multiplayer client also does all of this already and has worked fine until now.

tl;dr: In the name of simplicity, having all callbacks from online components be on the game thread is much easier and cleaner to handle. If this ever becomes a performance concern, such as if some operations can be done asynchronously, it should be a conscious decision by the author of the code and not by the online component.

### CODE REVIEW

File: osu.Game/Screens/Play/HUD/MultiplayerGameplayLeaderboard.cs [hunk 1]
```diff
@@ -80,6 +78,8 @@ protected override void LoadComplete()
             // BindableList handles binding in a really bad way (Clear then AddRange) so we need to do this manually..
             foreach (int userId in playingUsers)
             {
+                spectatorClient.WatchUser(userId);
```
Comment (peppy) [inline osu.Game/Screens/Play/HUD/MultiplayerGameplayLeaderboard.cs:81]: Curious about the moving of this call. Is there a reason it needs to be run from the update thread to guarantee behaviour? Or just felt like a better place for it to sit?

Comment (smoogipoo) [reply osu.Game/Screens/Play/HUD/MultiplayerGameplayLeaderboard.cs:81]: The other way is to schedule internally to `WatchUser`, otherwise it leads to potential cross-thread mutation of a list.

Comment (peppy) [reply osu.Game/Screens/Play/HUD/MultiplayerGameplayLeaderboard.cs:81]: Do we not want a bit more safety here? Can we use `ThreadSafety` checks at least?

Comment (smoogipoo) [reply osu.Game/Screens/Play/HUD/MultiplayerGameplayLeaderboard.cs:81]: Will add an assert, sure.

Comment (peppy) [reply osu.Game/Screens/Play/HUD/MultiplayerGameplayLeaderboard.cs:81]: Apart from this it looks fine. I'd probably add asserts anywhere the safety is required to play it safe.
